### PR TITLE
feat(models): Use CUDA graphs to accelerate spectral transforms for reduced grids

### PR DESCRIPTION
**Pull request by samhatfield**
## Description
This PR proposes to use graphs to accelerate the Fourier transform for reduced grids.

## What problem does this change solve?
#788 introduced spectral loss functions. This is accomplished by transforming gridded fields into spectral space, with a spherical harmonic basis. For reduced grids, this involves taking a Fourier transform (FFT) on each latitude ring from pole to pole. Since reduced grids have varying numbers of points per latitude, it's not possible to execute all latitudes in a single kernel like it is for regular grids. Instead it is accomplished with a loop. Each iteration of this loop executes a separate, small kernel on GPUs, so this is extremely inefficient.

Graphs are the best solution for such occasions, and essentially allow one to bundle all of the FFTs into a single kernel execution by recording the computational graph once and then replaying it on updated buffers. This PR contains the proposed graph implementation, started by myself with major contributions from @ssmmnn11. I will document progress here.

## What issue or task does this change relate to?
#599 

##  Additional notes ##
The earlier PR #933 was closed accidentally when its target branch was merged. However the branch in that PR did not implement a start-of-the-art graphed FFT so it needed to be updated anyway. This PR is the one that should be merged.

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

----
📚 Documentation preview 📚: https://anemoi-training--974.org.readthedocs.build/en/974/

----
📚 Documentation preview 📚: https://anemoi-graphs--974.org.readthedocs.build/en/974/

----
📚 Documentation preview 📚: https://anemoi-models--974.org.readthedocs.build/en/974/

### CODE REVIEW

File: models/tests/layers/test_sht.py [hunk 1]
```diff
@@ -130,3 +132,37 @@ def test_idempotency_inverse_direct(sht_setup):
         maxdiff = max(maxdiff, torch.abs((ref - got) / ref).max().item())
 
     assert maxdiff < tolerance
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize("sht_setup", ["reduced", "octahedral"], indirect=True)
+def test_multiple_direct_calls(sht_setup):
+    """Test direct transform can be called multiple times, to verify the CUDA graph functionality works correctly.
+    Reduced grids only.
```
[the remark this answers is not in the document]
Comment (samhatfield) [reply models/tests/layers/test_sht.py:139]: Octahedral grids are a kind of reduced grid.

File: models/tests/layers/test_sht.py [hunk 2]
```diff
@@ -130,3 +132,37 @@ def test_idempotency_inverse_direct(sht_setup):
...
+    dtype = sht_setup["dtype"]
+    direct = sht_setup["direct"]
+
+    before = torch.randn((1, 1, direct.n_grid_points), dtype=dtype)
+
+    once = direct(before)
+
+    twice = direct(before)
+
+    assert torch.all(once == twice)
```
[the remark this answers is not in the document]
Comment (samhatfield) [reply models/tests/layers/test_sht.py:150]: Nah.

File: models/src/anemoi/models/layers/spectral_helpers.py [hunk 3]
```diff
@@ -310,10 +367,10 @@ def __init__(self, lons_per_lat: list[int], truncation: int) -> None:
 
         # Use more efficient batched rfft for regular grids
         if len(set(self.lons_per_lat)) > 1:
-            LOGGER.info("InverseSphericalHarmonicTransform: Using irfft_rings_reduced for reduced grid")
+            LOGGER.debug("InverseSphericalHarmonicTransform: Using irfft_rings_reduced for reduced grid")
```
Comment (sahahner) [inline models/src/anemoi/models/layers/spectral_helpers.py]: I personally would leave these logger outputs as info (also above).

File: training/src/anemoi/training/losses/spectral.py [hunk 4]
```diff
@@ -109,12 +109,12 @@ def __init__(
             self.transform = DCT2D(**kwargs)
         elif transform == "reduced_sht":
             # expected additional args: grid
-            # optional args: truncation
+            # optional args: truncation, use_graphed_rfft
```
Comment (sahahner) [inline training/src/anemoi/training/losses/spectral.py:117]: A short comment about this optional argument in the loss function documentation [here](https://anemoi.readthedocs.io/projects/training/en/latest/modules/losses.html#example-configuration-reduced-gaussian-grid-sht) would be useful, I think.

**Comment by samhatfield** [comment]
Benchmark results with and without graphs, and also compared with ecTrans, all running on a single GH200:


Truncation | anemoi-core (graphs) | anemoi-core (no graphs) | ecTrans
-- | -- | -- | --
 79 | 2.4 | 19.7 | 3.1
159 | 5.4 | 40.7 | 8.5
319 | 12.1 | 89.6 | 26.5
639 | 34.7 | 206.8 | 92.9
1279 | 137.4 | 480.9 | 379.8

All reported times are the median wall time in ms to do one inverse followed by a direct transform, 400 fields. Benchmark script here:

<details>

<summary>benchmark.py</summary>

```python
import torch


def random_spectral_array(truncation: int, dtype: torch.dtype) -> torch.Tensor:
    # Shape: [batch index, ensemble member, l, m]
    shape = (1, 1, truncation + 1, truncation + 1)
    spectral_array = torch.complex(torch.randn(shape, dtype=dtype), torch.randn(shape, dtype=dtype))
    spectral_array[0, 0, :, 0].imag = 0.0  # m = 0 modes must be real

    # Zero the lower triangle, which has no meaning
    for i in range(truncation + 1):
        spectral_array[0, 0, :i, i] = 0.0 + 0.0j

    return spectral_array


def main(truncation: int, nfld: int, niter: int, use_graphs: bool):
    from anemoi.models.layers.spectral_helpers import InverseSphericalHarmonicTransform
    from anemoi.models.layers.spectral_helpers import SphericalHarmonicTransform
    from time import time

    device = "cuda"
    torch.set_default_device(device)
    dtype = torch.float32

    nlat = 2 * (truncation + 1)
    lons_per_lat = [20 + 4 * i for i in range(nlat // 2)]
    lons_per_lat += reversed(lons_per_lat)

    isht = InverseSphericalHarmonicTransform(
        lons_per_lat=lons_per_lat, truncation=truncation, use_graphed_irfft=use_graphs
    ).to(device)
    sht = SphericalHarmonicTransform(
        lons_per_lat=lons_per_lat, truncation=truncation, use_graphed_rfft=use_graphs
    ).to(device)

    original_spectral_array = random_spectral_array(truncation, dtype)
    spectral_array = original_spectral_array.clone()

    times = [0] * niter

    for i in range(niter):
        start_time = time()
        gridded = isht(spectral_array)
        torch.cuda.synchronize()
        spectral_array = sht(gridded)
        torch.cuda.synchronize()
        end_time = time()

        times[i] = 1000.0*(end_time - start_time)

        error = torch.max(torch.abs(spectral_array - original_spectral_array))

    print("truncation,nfld,min (ms),max (ms),med (ms)\n")
    print(f"{truncation},{nfld},{min(times):.6f},{max(times):.6f},{sorted(times)[len(times)//2]:.6f}\n")


if __name__ == "__main__":
    from argparse import ArgumentParser

    parser = ArgumentParser()
    parser.add_argument("--truncation", type=int, default=79, help="Spectral truncation")
    parser.add_argument("--nfld", type=int, default=1, help="Number of fields to transform")
    parser.add_argument("--niter", type=int, default=1, help="Number of iterations")
    parser.add_argument("--use_graphs", action="store_true", default=False, help="Whether to use torch.compile for the transforms")
    args = parser.parse_args()

    truncation = args.truncation
    nfld = args.nfld
    niter = args.niter
    use_graphs = args.use_graphs

    main(truncation, nfld, niter, use_graphs)

```

</details>

Don't read too much into the anemoi / ecTrans comparison -> for various reasons it's difficult to make a fair comparison. Still, it's nice to see that we're in the same ballpark as this very complicated Fortran library!

**Comment by samhatfield** [comment]
Thanks for the comments @sahahner. I think this is finally ready for merging, whenever you (+1 other?) approves it.

File: training/src/anemoi/training/losses/spectral.py (hunk 4 above)
Comment (samhatfield) [reply training/src/anemoi/training/losses/spectral.py:117]: I added a comment as a note.

**Comment by sahahner** [comment]
> Thanks for the comments @sahahner. I think this is finally ready for merging, whenever you (+1 other?) approves it.

Hi @samhatfield, sorry for being slow on this one. Would it be possible for you to resolve/go through the Copilot review comments, please? While there are unresolved comments, this cannot be merged.

File: models/tests/layers/test_sht.py [hunk 5]
```diff
@@ -130,3 +132,37 @@ def test_idempotency_inverse_direct(sht_setup):
...
+@pytest.mark.parametrize("sht_setup", ["reduced", "octahedral"], indirect=True)
+def test_direct_autograd_with_graphed_reduced_fft(sht_setup):
+    """Check gradients still work for reduced-grid FFT with CUDA graphs on."""
+    dtype = sht_setup["dtype"]
+    direct = sht_setup["direct"]
+
+    x = torch.randn((2, 1, direct.n_grid_points), dtype=dtype, requires_grad=True)
+    y = direct(x)
+    loss = torch.square(torch.abs(y)).mean()
+    loss.backward()
```
[the remark this answers is not in the document]
Comment (samhatfield) [reply models/tests/layers/test_sht.py:163]: It's a good point, but I'm not sure if it's really necessary. @sahahner do you know at what precision e.g. AIFS is trained?

File: models/src/anemoi/models/layers/spectral_helpers.py [hunk 6]
```diff
@@ -217,6 +228,58 @@ def rfft_rings_reduced(self, x: Tensor) -> Tensor:
...
+            LOGGER.info(f"Compiling graphed callable for rfft_rings_reduced with input signature {key}")
+            sample_x = torch.zeros_like(x, requires_grad=x.requires_grad)
+            with torch.amp.autocast("cuda", cache_enabled=False):
+                self._graphed_rfft_cache[key] = make_graphed_callables(
+                    tuple(
+                        partial(self.rfft_rings_reduced_banded, start_lat=latitude_band[0], end_lat=latitude_band[1])
+                        for latitude_band in self.latitude_bands
+                    ),
+                    tuple([(sample_x,)] * len(self.latitude_bands)),
+                )
```
[the remark this answers is not in the document]
Comment (samhatfield) [reply models/src/anemoi/models/layers/spectral_helpers.py:297]: Duplicate of comment above.

File: models/src/anemoi/models/layers/spectral_helpers.py [hunk 7]
```diff
@@ -217,6 +221,38 @@ def rfft_rings_reduced(self, x: Tensor) -> Tensor:
...
+        """
+
+        if x.device.type != "cuda":
+            return self.rfft_rings_reduced_naive(x)
+
+        key = (tuple(x.shape), x.dtype, x.device, x.requires_grad)
+        if key not in self._graphed_rfft_cache:
+            LOGGER.info(f"Compiling graphed callable for rfft_rings_reduced with input signature {key}")
+            sample_x = torch.zeros_like(x, requires_grad=x.requires_grad)
+            self._graphed_rfft_cache[key] = make_graphed_callables(self._rfft_reduced_impl, (sample_x,))
```
[the remark this answers is not in the document]
Comment (samhatfield) [reply models/src/anemoi/models/layers/spectral_helpers.py]: Copilot is not wrong here, but it's difficult for me to say whether this is really a big deal. My guess is that during training it is likely that the spectral transform will only encounter one unique tensor shape. For simplicity I suggest we ignore this for now, but if we encounter memory problems later, then we consider adding a cache eviction mechanism to control CUDA graph memory usage.

File: models/src/anemoi/models/layers/spectral_helpers.py [hunk 8]
```diff
@@ -217,6 +228,58 @@ def rfft_rings_reduced(self, x: Tensor) -> Tensor:
...
 
+    def rfft_rings_reduced_banded(self, x: Tensor, start_lat: int, end_lat: int) -> Tensor:
+        # Prepare zero-padded output tensor for filling with rfft
+        output_tensor = torch.zeros(
+            *x.shape[:-1],
+            end_lat - start_lat,
+            max(self.lons_per_lat) // 2 + 1,
+            device=x.device,
+            dtype=torch.complex64 if x.dtype == torch.float32 else torch.complex128,
+        )
```
[the remark this answers is not in the document]
Comment (samhatfield) [reply models/src/anemoi/models/layers/spectral_helpers.py]: Resolved.

File: models/src/anemoi/models/layers/spectral_helpers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ssmmnn11) [reply models/src/anemoi/models/layers/spectral_helpers.py]: Yes, you can ignore this. should it become a problem we can implement it.

File: models/tests/layers/test_sht.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ssmmnn11) [reply models/tests/layers/test_sht.py:163]: The loss comp. should always run in float32 - autocast is now off by default. We can add this later as well.

**Review by japols — APPROVED** [review]
Looks good to me. I guess we can make the magic `number_of_latitude_bands = 3` configurable if required in a follow-up PR.
